src/libpmemobj/tx.c

```diff
@@ -1013,8 +1019,16 @@ pmemobj_tx_commit(void)
 		operation_start(tx->lane->external);
 
 		struct user_buffer_def *userbuf;
-		VEC_FOREACH_BY_PTR(userbuf, &tx->redo_userbufs)
-			operation_add_user_buffer(tx->lane->external, userbuf);
+		struct operation_context *ctx = tx->lane->external;
```

Comment:
Like I indicated earlier in the issue, I don't think we need to check for this error here. Instead, we simply need to ensure that the `operation_add_user_buffer` never fails (like you've already done in `tx_construct_user_buffer`).

To have the FATAL/ASSERT that you added you can change the `operation_add_user_buffer` to return an error (-1) if `VEC_PUSH_BACK(&ctx->next, buffer_offset);` fails. This would let you avoid making all the memops structs public.

Reply:
not sure to fully follow your reasoning, but anyway I will try to push a new commit based on what I think I have understood...